src/hotspot/share/gc/shared/c2/barrierSetC2.cpp

```diff
@@ -699,127 +699,62 @@ Node* BarrierSetC2::obj_allocate(PhaseMacroExpand* macro, Node* mem, Node* toobi
                                  Node*& i_o, Node*& needgc_ctrl,
                                  Node*& fast_oop_ctrl, Node*& fast_oop_rawmem,
                                  intx prefetch_lines) const {
+  assert(UseTLAB, "Only for TLAB enabled allocations");
 
-  Node* eden_top_adr;
-  Node* eden_end_adr;
+  Node* thread = macro->transform_later(new ThreadLocalNode());
+  Node* tlab_top_adr = macro->basic_plus_adr(macro->top()/*not oop*/, thread, in_bytes(JavaThread::tlab_top_offset()));
+  Node* tlab_end_adr = macro->basic_plus_adr(macro->top()/*not oop*/, thread, in_bytes(JavaThread::tlab_end_offset()));
 
-  macro->set_eden_pointers(eden_top_adr, eden_end_adr);
-
-  // Load Eden::end.  Loop invariant and hoisted.
+  // Load TLAB end.
   //
-  // Note: We set the control input on "eden_end" and "old_eden_top" when using
-  //       a TLAB to work around a bug where these values were being moved across
+  // Note: We set the control input on "tlab_end" and "old_tlab_top" to work around
+  //       a bug where these values were being moved across
   //       a safepoint.  These are not oops, so they cannot be include in the oop
   //       map, but they can be changed by a GC.   The proper way to fix this would
   //       be to set the raw memory state when generating a  SafepointNode.  However
   //       this will require extensive changes to the loop optimization in order to
   //       prevent a degradation of the optimization.
   //       See comment in memnode.hpp, around line 227 in class LoadPNode.
-  Node *eden_end = macro->make_load(toobig_false, mem, eden_end_adr, 0, TypeRawPtr::BOTTOM, T_ADDRESS);
-
-  // We need a Region for the loop-back contended case.
-  enum { fall_in_path = 1, contended_loopback_path = 2 };
-  Node *contended_region;
-  Node *contended_phi_rawmem;
-  if (UseTLAB) {
-    contended_region = toobig_false;
-    contended_phi_rawmem = mem;
-  } else {
-    contended_region = new RegionNode(3);
-    contended_phi_rawmem = new PhiNode(contended_region, Type::MEMORY, TypeRawPtr::BOTTOM);
-    // Now handle the passing-too-big test.  We fall into the contended
-    // loop-back merge point.
-    contended_region    ->init_req(fall_in_path, toobig_false);
-    contended_phi_rawmem->init_req(fall_in_path, mem);
-    macro->transform_later(contended_region);
-    macro->transform_later(contended_phi_rawmem);
-  }
+  Node *tlab_end = macro->make_load(toobig_false, mem, tlab_end_adr, 0, TypeRawPtr::BOTTOM, T_ADDRESS);
```

Comment:
```suggestion
  Node* tlab_end = macro->make_load(toobig_false, mem, tlab_end_adr, 0, TypeRawPtr::BOTTOM, T_ADDRESS);
```

Some more occurences in below changes/code.

Reply:
Right, I fixed those in new commit.